## Agenda:
* Introduction to Numpy Functions.
* Numpy Function and Methods.
* Examples.
## Introduction to Numpy Functions:

Functions:
* sum
* mean
* argmax
* argmin
* argsort
* cov
* where
* cumsum
* cumprod
* choose
* T
* sqrt

In [3]:
import numpy as np

### Sum:
* Sum of array elements over a given axis.
* If we don't pass axis informtion it sums all the elements in given array.

In [19]:
a = np.array([1,2,3,4,5])

In [20]:
np.sum(a=a)

15

In [21]:
a.shape

(5,)

In [46]:
b = np.arange(100).reshape(10,10)

In [31]:
b

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
       [50, 51, 52, 53, 54, 55, 56, 57, 58, 59],
       [60, 61, 62, 63, 64, 65, 66, 67, 68, 69],
       [70, 71, 72, 73, 74, 75, 76, 77, 78, 79],
       [80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
       [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]])

In [24]:
np.sum(a=b,axis=1)

array([ 45, 145, 245, 345, 445, 545, 645, 745, 845, 945])

* In above example it produced sum of each row since axis is 1.
* If we pass keepdims to True it retains actual data dimentions.

In [43]:
np.sum(a=b,axis=1,keepdims=True)

array([[ 45],
       [145],
       [245],
       [345],
       [445],
       [545],
       [645],
       [745],
       [845],
       [945]])

* If we set out param to  True it assigns output to given out array(in below example out_array).
* Note: out array dimension should match with expected output of np.sum.

In [39]:
output_array = np.empty(shape=(10,1))

In [41]:
np.sum(a=b,axis=1,keepdims=True,out=output_array)

array([[  45.],
       [ 145.],
       [ 245.],
       [ 345.],
       [ 445.],
       [ 545.],
       [ 645.],
       [ 745.],
       [ 845.],
       [ 945.]])

In [44]:
output_array

array([[  45.],
       [ 145.],
       [ 245.],
       [ 345.],
       [ 445.],
       [ 545.],
       [ 645.],
       [ 745.],
       [ 845.],
       [ 945.]])

### Mean:
* Compute the arithmetic mean along the specified axis.
* If we don't pass axis information, the average is taken over the flattened array by default.

In [50]:
np.mean(b)

49.5

* The above code returned average of values in b array.

In [56]:
np.mean(a=b,axis=1)

array([  4.5,  14.5,  24.5,  34.5,  44.5,  54.5,  64.5,  74.5,  84.5,  94.5])

* The above one returned average along the axis 1 that is average of each row.

In [57]:
np.mean(a=b,axis=0)

array([ 45.,  46.,  47.,  48.,  49.,  50.,  51.,  52.,  53.,  54.])

* The above one returned average along the axis 0 that is average of each columns.
* np.mean also accepts keepdims and out params like np.sum.
* see below examples.

In [61]:
np.mean(a=b,axis=1,keepdims=True)

array([[  4.5],
       [ 14.5],
       [ 24.5],
       [ 34.5],
       [ 44.5],
       [ 54.5],
       [ 64.5],
       [ 74.5],
       [ 84.5],
       [ 94.5]])

In [69]:
out_array = np.empty(shape=(10,1))

In [70]:
np.mean(a=b,axis=1,keepdims=True,out=out_array)

array([[  4.5],
       [ 14.5],
       [ 24.5],
       [ 34.5],
       [ 44.5],
       [ 54.5],
       [ 64.5],
       [ 74.5],
       [ 84.5],
       [ 94.5]])

In [72]:
out_array

array([[  4.5],
       [ 14.5],
       [ 24.5],
       [ 34.5],
       [ 44.5],
       [ 54.5],
       [ 64.5],
       [ 74.5],
       [ 84.5],
       [ 94.5]])

### Argmax:
* Returns the indices of the maximum values along an axis.
* If we don't specify axis information, the index is into the flattened array, otherwise along the specified axis.
* If multiple max values present in array it just returns first max element index, ignores others.

In [80]:
a = np.array([5,6,3,2,0])

In [118]:
np.argmax(a)

1

* Above one returned an index of max element in array a.

In [96]:
b = np.array([[1,9,3,2,0],[8,9,5,7,4]])

In [97]:
b

array([[1, 9, 3, 2, 0],
       [8, 9, 5, 7, 4]])

In [126]:
np.argmax(b)

1

* In above example max element 9 is presented 2 times, but argmax returned only the index of first 9 value(1).

In [128]:
np.argmax(a=b,axis=1)

array([1, 0], dtype=int64)

* Returned corresponding indeces for each row since axis is 1.<br/>
### Argmin:<br/>
* Returns the indices of the minimum values along an axis.
* If we don't specify axis information, the index is into the flattened array, otherwise along the specified axis.
* If multiple max values present in array it just returns first max element index, ignores others.

In [102]:
a = np.array([5,6,3,2,0])

In [105]:
np.argmin(a)

4

* Above one returned an index of min element(0) in array.

In [116]:
b = np.array([[1,9,3,2,7],[8,0,5,7,4]])

In [114]:
b

array([[1, 9, 3, 2, 7],
       [8, 0, 5, 7, 4]])

In [115]:
np.argmin(b)

6

* Considered data as flat array since axis information is not available.
*  See below example for argmax with axis information.

In [129]:
np.argmin(b,axis=1)

array([0, 1], dtype=int64)

* Returned corresponding indeces for each row since axis is 1.<br/>

### Argsort: 
* Returns the indices that would sort an array.
* Using kind param we can specify algorithm to sort the data(example: kind ='quicksort').

In [131]:
a = np.array([4,2,1,3,9])

In [134]:
np.argsort(a=a)

array([2, 1, 3, 0, 4], dtype=int64)

In [141]:
np.argsort(a=a,kind='heapsort')

array([2, 1, 3, 0, 4], dtype=int64)

In [149]:
b = np.array([[1,9,3,2,7],[8,0,5,7,4]])

In [150]:
b

array([[1, 9, 3, 2, 7],
       [8, 0, 5, 7, 4]])

In [147]:
np.argsort(a=b,kind='quicksort',axis=0)

array([[0, 1, 0, 0, 1],
       [1, 0, 1, 1, 0]], dtype=int64)

In [148]:
np.argsort(a=b,kind='quicksort',axis=1)

array([[0, 3, 2, 4, 1],
       [1, 4, 2, 3, 0]], dtype=int64)

* When axis is 0 it compares values vertically.
* When axis is 1 it compares values horizontally.<br/> 
### Cov:<br/>
* Estimates a covariance matrix for given data and weights
* Covariance indicates the level to which two variables vary together.

In [155]:
b = np.array([[1,9,3,2,7],[8,0,5,7,4]])

In [156]:
b

array([[1, 9, 3, 2, 7],
       [8, 0, 5, 7, 4]])

In [157]:
np.cov(b)

array([[ 11.8 , -10.15],
       [-10.15,   9.7 ]])

* In above example <br/>
cov(b(row1,row1)) = 11.8 <br/>
cov(b(row1,row2)) = -10.15<br/>
cov(b(row2,row1)) = -10.15<br/>
cov(b(row2,row1)) = 9.7 

### Where:
* Returns elements depending on given condition.

In [164]:
a = np.array([4,2,1,3,9])

In [167]:
np.where(a > 3)

(array([0, 4], dtype=int64),)

* Returned indices of data which satisfy the condition.
* To check the data see below example.

In [245]:
indices = np.where(a > 3)

In [246]:
a[indices]

array([4, 9])

* 4 and 9 are greater than 3 in array a.

In [163]:
b = np.array([[1,9,3,2,7],[8,0,5,7,4]])

In [161]:
b

array([[1, 9, 3, 2, 7],
       [8, 0, 5, 7, 4]])

In [177]:
indexes = np.where(b > 5)

In [179]:
b[indexes]

array([9, 7, 8, 7])

* Here, [9 7 8 7] is result of given condition.

### Cumsum: 
* Returns the cumulative sum of the elements along a given axis.

In [181]:
a = np.array([4,2,1,3,9])

In [183]:
np.cumsum(a)

array([ 4,  6,  7, 10, 19], dtype=int32)

In [197]:
b = np.array([[1,9,3,2,7],[8,0,5,7,4]])

In [202]:
np.cumsum(b,axis=1)

array([[ 1, 10, 13, 15, 22],
       [ 8,  8, 13, 20, 24]], dtype=int32)

* Here,values are summed horizontally(sum along the row).

In [204]:
np.cumsum(b,axis=0)

array([[ 1,  9,  3,  2,  7],
       [ 9,  9,  8,  9, 11]], dtype=int32)

* Here,values are summed vertically(sum along the column).

### Cumprod:
* Return the cumulative product of the elements along a given axis.

In [206]:
a = np.array([4,2,1,3,9])

In [208]:
np.cumprod(a)

array([  4,   8,   8,  24, 216], dtype=int32)

In [210]:
b = np.array([[1,9,3,2,7],[8,0,5,7,4]])

In [213]:
np.cumprod(b,axis=1)

array([[  1,   9,  27,  54, 378],
       [  8,   0,   0,   0,   0]], dtype=int32)

In [215]:
np.cumprod(b,axis=0)

array([[ 1,  9,  3,  2,  7],
       [ 8,  0, 15, 14, 28]], dtype=int32)

### Choose:
* Constructs an array from an index array and a set of arrays to choose from.
* See below example for better understanding.

In [217]:
choices = [4,5,6,7]

In [220]:
np.choose(a=[2,3,2], choices=choices)

array([6, 7, 6])

* It created a new array using given indices and given choice data.
* See below examples to understand how it performs on complex data.

In [222]:
choices = [[0, 1, 2, 3], [10, 11, 12, 13],[20, 21, 22, 23], [30, 31, 32, 33]]

In [224]:
np.choose([[1],[2]], choices=choices)

array([[10, 11, 12, 13],
       [20, 21, 22, 23]])

In [226]:
np.choose([[1]], choices=choices)[0][2]

12

### T:
* Returns transpose of ndarray.
* Unlike self.transpose, It only transposes data when dimension of array is greater than 2.

In [228]:
b = np.array([[1,9,3,2,7],[8,0,5,7,4]])

In [230]:
b.ndim

2

In [232]:
b.T

array([[1, 8],
       [9, 0],
       [3, 5],
       [2, 7],
       [7, 4]])

In [234]:
a = np.array([4,2,1,3,9])

In [236]:
a.T

array([4, 2, 1, 3, 9])

* Since array a has only one dimension it didn't gave transpose of array a.

### Sqrt:
* Returns the positive square-root of an array, element-wise.

In [238]:
a = np.array([4,2,1,3,9])

In [240]:
np.sqrt(a)

array([ 2.        ,  1.41421356,  1.        ,  1.73205081,  3.        ])

In [242]:
b = np.array([-4,3,6,-9])

In [244]:
np.sqrt(a)

array([ 2.        ,  1.41421356,  1.        ,  1.73205081,  3.        ])

* Note: np sqrt always returns a positive value.